src/main/java/org/jabref/gui/EntryFromIDViewModel.java

```diff
@@ -0,0 +1,228 @@
+import javafx.beans.property.SimpleObjectProperty;
+import javafx.beans.property.SimpleStringProperty;
+import javafx.beans.property.StringProperty;
+import javafx.collections.FXCollections;
+import javafx.concurrent.Task;
+import javafx.concurrent.Worker;
+
+import java.util.regex.Pattern;
+import java.util.HashMap;
+import  java.util.function.Predicate;
+
+import javafx.scene.control.Label;
+import javafx.scene.control.TextArea;
+import org.jabref.Globals;
+import org.jabref.gui.duplicationFinder.DuplicateResolverDialog;
+import org.jabref.logic.bibtex.DuplicateCheck;
+import org.jabref.logic.bibtexkeypattern.BibtexKeyGenerator;
+import org.jabref.logic.importer.FetcherException;
+import org.jabref.logic.importer.IdBasedFetcher;
+import org.jabref.logic.importer.NoFetcherFoundException;
+import org.jabref.logic.importer.WebFetchers;
+import org.jabref.logic.importer.fetcher.*;
+import org.jabref.logic.l10n.Localization;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.strings.StringUtil;
+import org.jabref.preferences.JabRefPreferences;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EntryFromIDViewModel {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EntryFromIDViewModel.class);
+
+    private final JabRefPreferences prefs;
+    private final BooleanProperty searchingProperty = new SimpleBooleanProperty();
+    private final BooleanProperty searchSuccesfulProperty = new SimpleBooleanProperty();
+    private final StringProperty idText = new SimpleStringProperty();
+    private  String fetcherName;
+    private final BooleanProperty focusAndSelectAllProperty = new SimpleBooleanProperty();
+    private Task<Optional<BibEntry>> fetcherWorker = new FetcherWorker();
+    private HashMap <Predicate<String>,IdBasedFetcher> fetcherMatchPattern = new HashMap<Predicate<String>,IdBasedFetcher>();
+    private final BasePanel basePanel;
+    private final DialogService dialogService;
+
+    public EntryFromIDViewModel(JabRefPreferences preferences, BasePanel basePanel, DialogService dialogService) {
+        this.basePanel = basePanel;
+        this.prefs = preferences;
+        this.dialogService = dialogService;
+
+        fetcherMatchPattern.putAll(WebFetchers.getHashMapPredicateIdBasedFetchers(preferences.getImportFormatPreferences()));
+
+    }
+
+    public BooleanProperty searchSuccesfulProperty() {
+        return searchSuccesfulProperty;
+    }
+
+    public BooleanProperty searchingProperty() {
+        return searchingProperty;
+    }
+
+    public StringProperty idTextProperty() {
+        return idText;
+    }
+
+    public BooleanProperty getFocusAndSelectAllProperty() {
+        return focusAndSelectAllProperty;
+    }
+
+
+    public void stopFetching() {
+        if (fetcherWorker.getState() == Worker.State.RUNNING) {
+            fetcherWorker.cancel(true);
+        }
+    }
+
+    private class FetcherWorker extends Task<Optional<BibEntry>> {
```

Comment:
Doesn't JabRef have an infrastructure for background tasks somehow? Maybe, this is a TODO. The current code here feels a bit quickly hacked. At least, this nested class should be factored out to a separate class.

Reply:
The `BackgroundTask` class is perfectly made for such cases.